substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,603 @@
+	for_tuples!( type Pre = ( #( Tuple::Pre ),* ); );
+
+	fn weight(&self, call: &Call) -> Weight {
+		let mut weight = Weight::zero();
+		for_tuples!( #( weight = weight.saturating_add(Tuple.weight(call)); )* );
+		weight
+	}
+
+	fn validate(
+		&self,
+		origin: <Call as Dispatchable>::RuntimeOrigin,
+		call: &Call,
+		info: &DispatchInfoOf<Call>,
+		len: usize,
+		self_implicit: Self::Implicit,
+		inherited_implication: &impl Encode,
+	) -> Result<
+		(ValidTransaction, Self::Val, <Call as Dispatchable>::RuntimeOrigin),
+		TransactionValidityError,
+	> {
+		let valid = ValidTransaction::default();
+		let val = ();
+		let following_explicit_implications = for_tuples!( ( #( &self.Tuple ),* ) );
+		let following_implicit_implications = self_implicit;
+
+		for_tuples!(#(
+			// Implication of this pipeline element not relevant for later items, so we pop it.
+			let (_item, following_explicit_implications) = following_explicit_implications.pop_front();
+			let (item_implicit, following_implicit_implications) = following_implicit_implications.pop_front();
+			let (item_valid, item_val, origin) = {
+				let implications = (
+					// The first is the implications born of the fact we return the mutated
+					// origin.
+					inherited_implication,
+					// This is the explicitly made implication born of the fact the new origin is
+					// passed into the next items in this pipeline-tuple.
+					&following_explicit_implications,
+					// This is the implicitly made implication born of the fact the new origin is
+					// passed into the next items in this pipeline-tuple.
+					&following_implicit_implications,
+				);
+				Tuple.validate(origin, call, info, len, item_implicit, &implications)?
+			};
+			let valid = valid.combine_with(item_valid);
+			let val = val.push_back(item_val);
+		)* );
+		Ok((valid, val, origin))
+	}
+
+	fn prepare(
+		self,
+		val: Self::Val,
+		origin: &<Call as Dispatchable>::RuntimeOrigin,
+		call: &Call,
+		info: &DispatchInfoOf<Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Ok(for_tuples!( ( #(
+			Tuple::prepare(self.Tuple, val.Tuple, origin, call, info, len)?
+		),* ) ))
+	}
+
+	fn post_dispatch_details(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<Call>,
+		post_info: &PostDispatchInfoOf<Call>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<Weight, TransactionValidityError> {
+		let mut total_unspent_weight = Weight::zero();
+		for_tuples!( #({
+			let unspent_weight = Tuple::post_dispatch_details(pre.Tuple, info, post_info, len, result)?;
+			total_unspent_weight = total_unspent_weight.saturating_add(unspent_weight);
+		})* );
+		Ok(total_unspent_weight)
+	}
+
+	fn post_dispatch(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<Call>,
+		post_info: &mut PostDispatchInfoOf<Call>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		for_tuples!( #( Tuple::post_dispatch(pre.Tuple, info, post_info, len, result)?; )* );
+		Ok(())
+	}
```

Comment:
```suggestion
```

We don't need this, as we have `post_dispatch_details` overwritten correctly.

Reply:
We do because we need the `post_dispatch` function to correctly carry over the refunded weight from extension to extension in the pipeline. If any extension in the pipeline refunds some weight, we want all of the other extensions' post dispatch logic to see the change.

There are tests for this in `substrate/frame/support/src/dispatch.rs` in the `extension_weight_tests` that test this exact behavior. Those tests will fail if you try to remove the `post_dispatch` tuple impl.